File: clojure/resistor-color-trio/src/resistor_color_trio.clj
(ns resistor-color-trio)

(def color->code {"black" 0 "brown" 1 "red" 2 "orange" 3 "yellow" 4
                  "green" 5 "blue" 6 "violet" 7 "grey" 8 "white" 9})

(def prefixes [[1000000000 "giga"] [1000000 "mega"] [1000 "kilo"]])

(defn resistor-label
  "Returns the resistor label based on the given color bands."
  [colors]
  (let [[tens-band units-band multiplier-band] colors
        base (+ (* 10 (color->code tens-band)) (color->code units-band))
        multiplier (reduce * (repeat (color->code multiplier-band) 10))                
        ohms (* base multiplier)]
    (if (zero? ohms) 
      "0 ohms"
      (if-let [[scale prefix] (some (fn [[scale prefix]]
                                      (when (zero? (mod ohms scale))
                                        [scale prefix]))
                                    prefixes)]
        (str (/ ohms scale) " " prefix "ohms")
        (str ohms " ohms")))))

File: python/resistor-color-trio/resistor_color_trio.py
"""Convert a 3-band resistor to its value with a label"""

COLORS = {
    'black': 0, 'brown': 1, 'red': 2, 'orange': 3, 'yellow': 4,
    'green': 5, 'blue': 6, 'violet': 7, 'grey': 8, 'white': 9
}

VALUE_PREFIXES = {1_000_000_000: 'giga', 1_000_000: 'mega', 1_000: 'kilo'}

def label(colors):
    tens, units, exponent, *_ = [COLORS[color] for color in colors]
    ohms = (10 * tens + units) * 10 ** exponent

    for value, prefix in VALUE_PREFIXES.items():
        if ohms > value:
            return f'{ohms // value} {prefix}ohms'

    return f'{ohms} ohms'
    